### PR TITLE
monitor: check the on-demand device databases of a release

### PR DESCRIPTION
**Pull request by cavearr**
Follows up on #947: you suggested adding the openxc7 chipdb verification to `monitor-remote-configs`, and this is it. One file, `scripts/check_remote_configs.py`; the workflow itself is unchanged.

**Why.** For openxc7 "the release exists and is stable" no longer means apio can build: since the on-demand model the packages carry no device databases, and apio downloads the one a board needs from a per-part release asset that `PARTS-INDEX.json` names. If one of those assets is missing or was published truncated, apio is dead for that FPGA and nothing notices until a user hits it — the monitor is the natural place to catch it.

**What it checks**, for any release that publishes a parts index (packages without one are untouched, so this is a no-op everywhere else):

- the index is the one named after the release's tag (apio's date rule);
- its `schema` is the one `apio/managers/xilinx_chipdb.py` reads — a bump means the fields the loader uses were renamed, and the message says so;
- its `release-tag` is this release (an index from another release would send apio to another release's assets);
- `part-count`/`generated-count`/`chipdb-count` match what the document actually describes;
- every part that has a database has its asset **published, at the size the index promises**.

**Cost.** Sizes come from the release metadata the script already fetches, so the only extra request is the index itself (a few tens of KB) — and the asset loop deduplicates by file, because the speed grades of a part share one database: 15 checks instead of 56 for today's openxc7 release. Byte-level (sha256) verification stays where it is cheap to do once, in the toolchain repo's promotion gate.

**Verified against the live configs** (`GITHUB_TOKEN` set, full run green, exit 0):

```
===== Remote Config apio-1.5.x.jsonc =====
Checking package [openxc7] release [2026-08-20]          <- predates the model, no index, skipped
===== Remote Config apio-1.6.x.jsonc =====
Checking package [openxc7] release [2026-08-31]
Checking parts index [apio-xilinx-parts-index-20260831.json]
Parts index OK (154 parts, 56 of them with a database, in 15 assets)
Done (Check OK 5 remote configs).
```

And the failure paths, exercised with a doctored index so they are not just theory:

```
asset missing  : exit 1 · Error: part 'xc7a35tcsg324-1' needs asset 'a.bin.tgz', which this release does not publish
asset truncated: exit 1 · Error: asset 'a.bin.tgz' is 99 bytes, but the index says 100
bad counts     : exit 1 · Error: i.json: generated-count is 7, but the document describes 1
foreign index  : exit 1 · Error: expected index 'apio-xilinx-parts-index-20260831.json', found '…-20260830.json'
```

Lint: `black`, `flake8` and `pylint` at the `tox.ini` versions — clean, pylint 10.00/10.

One judgement call worth your opinion: a schema bump on our side makes this monitor **red**, deliberately — that is exactly the case where apio's loader would silently stop finding the sizes. If you would rather it warned instead of failing, say so and I will change it. I did not add tests: `scripts/` has none today and I did not want to bring a new layout in with this; happy to add them if you want them.

### CODE REVIEW

**Comment by zapta** [comment]
Thanks @cavearr!

1. I noticed that the index file in the release is different than PARTS-INDEX.json in the package itself. How about replacing it with PARTS-INDEX.json to be consistent across the package?  (Maybe even just the name 'PARTS-INDEX.json' without the tag, similar to SHA256SUMS, since it contains the tags internally).

image


2. By the way, why SHA256SUMS contains the checksums just of the package assets and not the chipdb assets?

3. You said "for any release that publishes a parts index". How about changing the criteria to "apio version >= 1.6". Otherwise, if the parts index is missing we will assume that the release is ok.

I will now review the code and add there comments if any.

File: scripts/check_remote_configs.py [hunk 1]
```diff
@@ -25,6 +25,17 @@
...
+# -- Prefix of the parts index asset published by packages whose device
+# -- databases are fetched on demand (openxc7). The rest of the asset name
+# -- is the release tag's date, the same rule apio uses for every asset.
+PARTS_INDEX_PREFIX = "apio-xilinx-parts-index-"
+
+# -- The parts index schema that apio's loader
+# -- (apio/managers/xilinx_chipdb.py) knows how to read. A release with a
+# -- different schema renamed or reshaped the fields the loader uses, so
+# -- bump this together with the loader.
+PARTS_INDEX_SCHEMA = 5
```
Comment (zapta) [inline scripts/check_remote_configs.py]: Add "_VERSION" for clarity?

File: scripts/check_remote_configs.py [hunk 2]
```diff
@@ -84,12 +95,126 @@ def check_package(package_name: str, package_config: Dict):
...
-    for asset_name in asset_names:
+    assets = {asset["name"]: asset for asset in data["assets"]}
+    for asset_name in assets:
         print(f"- {asset_name}")
 
     print("Release exists and is stable")
 
+    # -- If this package fetches its device databases on demand, check that
+    # -- the databases its index promises are actually published.
+    check_parts_index(tag, assets)
```
Comment (zapta) [inline scripts/check_remote_configs.py]: Shouldn't we call this only if the package is openxc7 since the the part index json is specific for openxc7?

File: scripts/check_remote_configs.py [hunk 3]
```diff
@@ -84,12 +95,126 @@ def check_package(package_name: str, package_config: Dict):
...
+        return
+    assert len(index_names) == 1, index_names
+    index_name = index_names[0]
+
+    print()
+    print(f"Checking parts index [{index_name}]")
+
+    # -- Asset names are derived from the tag's date, so the index of this
+    # -- release must be the one named after this tag.
+    expected_name = PARTS_INDEX_PREFIX + tag.replace("-", "") + ".json"
```
Comment (zapta) [inline scripts/check_remote_configs.py]: I suggest to call the file PARTS-INDEX.json, with no tag, similar to SHA256SUMS.

File: scripts/check_remote_configs.py [hunk 4]
```diff
@@ -84,12 +95,126 @@ def check_package(package_name: str, package_config: Dict):
...
+            )
+            sys.exit(1)
+
+    print(
+        f"Parts index OK ({len(parts)} parts, {len(generated)} of them "
+        f"with a database, in {len(checked)} assets)"
+    )
+
+
+def check_parts_index(tag: str, assets: Dict):
```
Comment (zapta) [inline scripts/check_remote_configs.py]: add 'openxc7' to the name since the check is specific to openxc7.

File: scripts/check_remote_configs.py [hunk 5]
```diff
@@ -84,12 +95,126 @@ def check_package(package_name: str, package_config: Dict):
...
         print(f"- {asset_name}")
 
     print("Release exists and is stable")
 
+    # -- If this package fetches its device databases on demand, check that
+    # -- the databases its index promises are actually published.
+    check_parts_index(tag, assets)
+
+
+def check_parts_index_content(index_name: str, index: Dict, assets: Dict):
```
Comment (zapta) [inline scripts/check_remote_configs.py]: add 'openxc7' to the name since the check is specific to openxc7.

**Comment by cavearr** [comment]
Thanks @zapta!! all four are right. Pushed the two that live in this script; the other two are changes on the toolchain side and I have them queued there.

**1. Asset name.** Agreed, and it is better than what we publish today: one document, one name, and the tag is inside it anyway (`release-tag`), which is the check that actually catches a release built across midnight — the name was only a proxy for it. The toolchain will publish it as `PARTS-INDEX.json`, like `SHA256SUMS`. So that the two repos do not have to change in the same minute, this script now accepts **either** name (the dated one still has to be the one named after its tag), so you can merge whenever and the rename lands on its own.

**2. `SHA256SUMS` only covering the packages** is history, not design: it predates the on-demand model, when the packages were the only assets. The database checksums do exist — `asset-sha256` (the download) and `chipdb-sha256` (the file on disk) in the index, which is where the loader needs them per part — but you are right that the human path (`sha256sum -c SHA256SUMS`) should cover everything the release publishes. I will extend it to all assets, generated in the same job from the same bytes as the index, so the two cannot drift.

**3. Missing index.** Good catch, that was a real hole: "no index → skip" is right for the old model and wrong for a broken upload. Fixed with an explicit threshold rather than an apio version, because what changed is a property of the package: `PARTS_INDEX_REQUIRED_FROM = {"openxc7": "2026-08-29"}` — from that tag on the packages carry no databases, so a release without an index is broken and now exits 1 naming the package and the threshold; earlier releases (what 1.5.x points at) are still skipped as legacy. If you prefer it keyed by config version instead, say so and I will change it — the version is in the file name, so it is a small edit.

**4. `_VERSION`** — done, `PARTS_INDEX_SCHEMA_VERSION`.

Verified after the change: the live configs still pass (1.6.x → "Parts index OK (154 parts, 56 of them with a database, in 15 assets)", 1.5.x → openxc7 2026-08-20 skipped as legacy), an openxc7 release after the threshold without an index exits 1, and both index names resolve. Lint clean at the `tox.ini` versions.

**Comment by zapta** [comment]
Hi @cavearr, all sounds good.

One more thing, the chipdb assert verification will be performed only for the openxc7, right? That is, the workflow will be hard coded to expect PARTS-INDEX and verify it only if the package is 'openxc7', right?

(We can generalize it in the future but I think for now the design of the on-demand is  xilinx chipdb specific).

**Comment by cavearr** [comment]
@zapta 

Yes,  and done, all four comments. The check now runs only for `openxc7`, and everything it touches says so: `OPENXC7_PACKAGE`, `OPENXC7_PARTS_INDEX_NAME` / `_PREFIX` / `_REQUIRED_FROM` / `_SCHEMA_VERSION`, `check_openxc7_parts_index()`, `find_openxc7_parts_index()`. The per-package threshold map became a single constant now that there is one package in it, with a comment marking it as the place to generalize from if a second one ever fetches something on demand. You are right that guessing at that generalization now would only be speculative.

Behaviour is unchanged for openxc7: the live configs still report `Parts index OK (154 parts, 56 of them with a database, in 15 assets)` for 1.6.x, and 1.5.x's `2026-08-20` is still skipped as legacy. Lint clean at the `tox.ini` versions.

**Your third comment is already live on the toolchain side.** Since this morning's release the index is published as `PARTS-INDEX.json`, exactly as you suggested — same name inside the package and in the release, like `SHA256SUMS`. And `SHA256SUMS` now covers every asset (your point 2 from the other day): `2026-09-01` carries 19 lines — three packages, the index and the 15 chipdb assets — generated in the same job from the same bytes as the index, so the two documents cannot drift. I checked them against each other after publication: every asset the index names has a line, with the same sha256 it declares, and no line describes anything the index does not. Nothing to do on your side: the script here accepts both names, so releases up to `2026-08-31` keep verifying too.

And thanks for #992/#993/#995 — that is the three loader observations from the other day implemented (schema check, sha256 of both the download and the unpacked file, and the leftover `.tgz` removed). Nice to see them land that fast.
